python/nano/src/bigdl/nano/pytorch/plugins/ddp_spawn.py

```diff
@@ -33,188 +33,241 @@
+from pytorch_lightning.core.datamodule import LightningDataModule
+from pytorch_lightning.strategies.launchers import _SpawnLauncher
+from pytorch_lightning.strategies import Strategy, DDPSpawnStrategy as _DDPSpawnStrategy
 from pytorch_lightning.plugins.environments import LightningEnvironment
-from pytorch_lightning.utilities.distributed import rank_zero_only
-from pytorch_lightning.utilities.seed import reset_seed
+from pytorch_lightning.overrides import LightningDistributedModule
+from pytorch_lightning.utilities.optimizer import optimizers_to_device
+from pytorch_lightning.utilities.distributed import ReduceOp
 
 from bigdl.nano.common.cpu_schedule import schedule_workers
-from bigdl.nano.deps.ipex.ipex_api import ipex_device, ipex_optimize
-import logging
+from bigdl.nano.deps.ipex.ipex_api import ipex_device, ipex_optimize, \
+    create_IPEXAccelerator_1_9, to_cpu
+from bigdl.nano.pytorch.utils import TORCH_VERSION_LESS_1_10
 
+import logging
 import warnings
 
+_STEP_OUTPUT_TYPE = Union[torch.Tensor, Dict[str, Any]]
+
 log = logging.getLogger(__name__)
 
 
-def start_processes_new(fn, args=(), nprocs=1, join=True, daemon=False,
-                        start_method='spawn', cpu_procs=None):
-    """Start processess with optimized environment variables."""
-    mp = multiprocessing.get_context(start_method)
-    error_queues = []
-    processes = []
-
-    if cpu_procs is None:
-        cpu_procs = schedule_workers(nprocs)
-
-    init_KMP_AFFINITY = os.environ.get("KMP_AFFINITY", "")
-    init_OMP_NUM_THREADS = os.environ.get("OMP_NUM_THREADS", "")
-
-    for i in range(nprocs):
-        os.environ["KMP_AFFINITY"] = f"granularity=fine,proclist"\
-                                     f"=[{','.join([str(i) for i in cpu_procs[i]])}],explicit"
-        os.environ["OMP_NUM_THREADS"] = str(len(cpu_procs[i]))
-        log.debug(f"[Process {i}]: using KMP_AFFINITY: {os.environ['KMP_AFFINITY']}")
-        log.debug(f"[Process {i}]: using OMP_NUM_THREADS: {os.environ['OMP_NUM_THREADS']}")
-        error_queue = mp.SimpleQueue()
-        process = mp.Process(
-            target=_wrap,
-            args=(fn, i, args, error_queue),
-            daemon=daemon,
-        )
-        process.start()
-        error_queues.append(error_queue)
-        processes.append(process)
+class _DDPSpawnLauncher(_SpawnLauncher):
+
+    def __init__(self, strategy: Strategy) -> None:
+        super().__init__(strategy)
+
+    def launch(self, function: Callable, *args: Any,
+               trainer: Optional["pl.Trainer"] = None, **kwargs: Any) -> Any:
+        # pytorch_lightning 1.6 uses this method to create child processes
+        os.environ["MASTER_PORT"] = str(self._strategy.cluster_environment.main_port)
+
+        if self._strategy.cpu_for_each_process is None:
+            cpu_procs = schedule_workers(self._strategy.num_processes)
+        else:
+            cpu_procs = self._strategy.cpu_for_each_process
+
+        # fix bug
```

Comment:
maybe more explanation here? Fix which bug? And how do you fix this bug?

Reply:
ok, I'll add more explanation